graphite-ch-optimizer.go

```diff
@@ -0,0 +1,310 @@
+	fd.Uint("loop-interval", v.GetUint("daemon.loop-interval"), "Daemon will check if there partitions to merge once per this interval, seconds")
+	fd.BoolP("dry-run", "n", v.GetBool("daemon.dry-run"), "Will print how many partitions would be merged without actions")
+	fl.String("output", v.GetString("logging.output"), "The logs file. '-' is accepted as STDOUT")
+	fl.String("log-level", v.GetString("logging.level"), "Valid options are: panic, fatal, error, warn, warning, info, debug, trace")
+	pflag.CommandLine.AddFlagSet(fc)
+	pflag.CommandLine.AddFlagSet(fd)
+	pflag.CommandLine.AddFlagSet(fl)
+	pflag.Parse()
+	v.SetConfigFile(*customConfig)
+	v.ReadInConfig()
+	log.Trace(v.ConfigFileUsed())
+	if *printDefaults {
+		t, err := toml.TreeFromMap(v.AllSettings())
+		if err != nil {
+			log.Fatal(err)
+		}
+		fmt.Println(t.String())
+		os.Exit(0)
+	}
+	fc.VisitAll(func(f *pflag.Flag) {
+		v.BindPFlag("clickhouse."+f.Name, f)
+	})
+	fd.VisitAll(func(f *pflag.Flag) {
+		v.BindPFlag("daemon."+f.Name, f)
+	})
+	fl.VisitAll(func(f *pflag.Flag) {
+		v.BindPFlag("logging."+f.Name, f)
+	})
+	if v.GetBool("daemon.dry-run") {
+		v.Set("daemon.one-shot", true)
+	}
+	v.Unmarshal(c)
+	log.Traceln(c)
+	return c
+}
+
+func main() {
+	if cfg.Daemon.OneShot {
+		optimize(cfg)
+	} else {
+		var wg sync.WaitGroup
+		wg.Add(1)
+		go func() {
+			log.Trace("Starting loop function")
+			for {
+				err := optimize(cfg)
+				if err != nil {
+					log.Errorf("Optimization failed: %s", err)
+				}
+				time.Sleep(time.Second * time.Duration(cfg.Daemon.LoopInterval))
+			}
+		}()
+		wg.Wait()
+	}
+}
+
+func optimize(cfg *Config) error {
+	connect, err := sql.Open("clickhouse", cfg.ClickHouse.ServerDsn)
+	err = connect.Ping()
+	if checkErr(err) != nil {
+		log.Fatalf("Ping ClickHouse server failed: %v", err)
+	}
+	rows, err := connect.Query(
+		SelectUnmerged,
+		sql.Named("Interval", cfg.ClickHouse.OptimizeInterval),
+	)
+	if checkErr(err) != nil {
+		return err
+	}
+	columns, err := rows.Columns()
+	if checkErr(err) != nil {
+		return err
+	}
+	log.WithField("columns", columns).Debug("The columns in the query:")
+	merges := []merge{}
+	for rows.Next() {
+		var (
+			m          merge
+			age        uint64
+			parts      uint64
+			maxTime    time.Time
+			rollupTime time.Time
+			modifiedAt time.Time
+		)
+		err = rows.Scan(&m.table, &m.partitionID, &m.partitionName, &age, &parts, &maxTime, &rollupTime, &modifiedAt)
+		if checkErr(err) != nil {
+			return err
+		}
+		merges = append(merges, m)
+		log.WithFields(log.Fields{
+			"table":          m.table,
+			"partition_id":   m.partitionID,
+			"partition_name": m.partitionName,
+			"age":            age,
+			"parts":          parts,
+			"max_time":       maxTime,
+			"rollup_time":    rollupTime,
+			"modified_at":    modifiedAt,
+		}).Debug("Merge to be applied")
+	}
+	log.Infof("Merges will be applied: %d", len(merges))
+	if cfg.Daemon.DryRun {
```

Comment:
wouldn't it be a bit nicer if we don't log `Merges will be applied` in dry mode and instead something like `Merges would be applied`? because now we get 2 messages in dry mode that contradict each other

Reply:
I like it, yes